src/Models/HydrostaticFreeSurfaceModels/implicit_free_surface.jl

```diff
@@ -1,78 +1,128 @@
-    vertically_integrated_lateral_face_areas = (Ax = Ax_zintegral, Ay=Ay_zintegral)
+    # Initialize vertically integrated lateral face areas
+    ∫ᶻ_Axᶠᶜᶜ = ReducedField(Face, Center, Nothing, arch, grid; dims=3)
+    ∫ᶻ_Ayᶜᶠᶜ = ReducedField(Center, Face, Nothing, arch, grid; dims=3)
+
+    vertically_integrated_lateral_face_areas = (xᶠᶜᶜ = ∫ᶻ_Axᶠᶜᶜ, yᶜᶠᶜ = ∫ᶻ_Ayᶜᶠᶜ)
+
     compute_vertically_integrated_lateral_face_areas!(vertically_integrated_lateral_face_areas, grid, arch)
 
-    ## implicit_step_solver = ImplicitFreeSurfaceSolver(arch, η, vertically_integrated_lateral_face_areas; maxit=100)
-    implicit_step_solver = ImplicitFreeSurfaceSolver(arch, η, vertically_integrated_lateral_face_areas)
+    implicit_step_solver = PreconditionedConjugateGradientSolver(implicit_free_surface_linear_operation!,
+                                                                 template_field = η,
+                                                                 maximum_iterations = grid.Nx * grid.Ny,
+                                                                 free_surface.solver_settings...)
+    
+    implicit_step_right_hand_side = ReducedField(Center, Center, Nothing, arch, grid; dims=3)
 
-    return ImplicitFreeSurface(η, g, 
-                               barotropic_volume_flux, 
+    return ImplicitFreeSurface(η,
+                               g,
+                               barotropic_volume_flux,
                                vertically_integrated_lateral_face_areas,
-                               implicit_step_solver)
+                               implicit_step_right_hand_side,
+                               implicit_step_solver,
+                               free_surface.solver_settings)
 end
 
-### In final form the two functions below will return 0 (these are invoked in hydrostatic_free_surface_tendency_kernel_functions.jl )
-explicit_barotropic_pressure_x_gradient(i, j, k, grid, free_surface::ImplicitFreeSurface) = 0
+@inline explicit_barotropic_pressure_x_gradient(i, j, k, grid, ::ImplicitFreeSurface) = 0
+@inline explicit_barotropic_pressure_y_gradient(i, j, k, grid, ::ImplicitFreeSurface) = 0
+
+"""
+Implicitly step forward η.
+"""
+ab2_step_free_surface!(free_surface::ImplicitFreeSurface, velocities_update, model, Δt, χ) =
+    implicit_free_surface_step!(free_surface::ImplicitFreeSurface, velocities_update, model, Δt, χ)
+
+function implicit_free_surface_step!(free_surface::ImplicitFreeSurface, velocities_update, model, Δt, χ)
+
+    η = free_surface.η
+    g = free_surface.gravitational_acceleration
+    rhs = free_surface.implicit_step_right_hand_side
+    ∫ᶻ_Q = free_surface.barotropic_volume_flux
+    ∫ᶻ_A = free_surface.vertically_integrated_lateral_face_areas
+
+    #=
+    # Take an explicit step first to produce an improved initial guess for η for the iterative solver.
+    event = explicit_ab2_step_free_surface!(free_surface, nothing, model, Δt, χ)
+    wait(device(model.architecture), event)
+    =#
```

Comment:
Should these comments be removed? If it's useful to take an explicit step first, maybe this can be controlled via a keyword argument?

Reply:
Ooh, that's a nice alternative to a comment :-D